js/ui/appActivation.js

```diff
@@ -275,17 +269,15 @@ const AppActivationContext = new Lang.Class({
         if (this._isBogusWindow(app))
             return;
 
-        appSystem.disconnect(this._appStateId);
-        this._appStateId = 0;
-
-        let aborted = this._abort;
-        this._abort = false;
 
-        if (aborted) {
+        if (this._abort) {
             this._app.request_quit();
             this._clearSplash();
+            this._abort = false;
         } else if (app.state == Shell.AppState.STOPPED) {
-            this._clearSplash();
+            // User cancelled the splash screen, but we need to wait
+            // for the running state before requesting to quit the app.
+            this._abort = true;
```

Comment:
I don't completely follow the flow here... If we get `STOPPED` for a given app, how can we get another notification later for the `RUNNING` state?

If what you see is a transition like this `STARTING->(speedwagon close)->STOPPED->(app window is created)->RUNNING->(request quit)->STOPPED`, could we further clean that up and make it so that if `STOPPED` is reached from the `STARTING` state, we don't forward that out and handle the quit request directly `ShellAppSystem`?

From the outside, I think we should only see `STARTING->STOPPED` or `STARTING->RUNNING->STOPPED` state transitions sequences.

Reply:
> I don't completely follow the flow here... If we get STOPPED for a given app, how can we get another notification later for the RUNNING state?

As the code in shell-app.c is right now, you can get into this situation if you explicitly close the Speedwagon window, which 99% of the times will happen before any "interesting window" for the app itself is created, thus moving to STOPPED.

However, the app itself is still not killed, since `meta_window_delete()` could not be called yet (no windows => no-op), so the app will eventually get into the RUNNING state after that point, which is when we call `request_quit()`.

> If what you see is a transition like this STARTING->(speedwagon close)->STOPPED->(app window is created)->RUNNING->(request quit)->STOPPED, could we further clean that up and make it so that if STOPPED is reached from the STARTING state, we don't forward that out and handle the quit request directly ShellAppSystem?

That's the other possibility I mentioned in the ticket, which would involve killing (e.g. SIGTERM) the process being launched at the time the speedwagon is closed. That would involve 2 different code paths, though, once for apps launched via D-Bus (e.g. totem) and one for apps spawned/forked directly (e.g. Chromium), since in each case there are different ways to get the PID.

Since all this code was already quite messy I proposed to have a simpler solution for now which is based on what we were trying to attempt already (deferring quitting the app until it's RUNNING), as that is a single code path, so that this doesn't get much in the way of the rebase.

If you think it's important enough to do the right thing or do nothing, then I'd suggest waiting until the rebase is done and do it afterwards. Thoughts?

> From the outside, I think we should only see STARTING->STOPPED or STARTING->RUNNING->STOPPED state transitions sequences.

I agree with this as being the right thing to do for the long term.

Reply:
>That's the other possibility I mentioned in the ticket, which would involve killing (e.g. SIGTERM) the process being launched at the time the speedwagon is closed. That would involve 2 different code paths, though, once for apps launched via D-Bus (e.g. totem) and one for apps spawned/forked directly (e.g. Chromium), since in each case there are different ways to get the PID.

>Since all this code was already quite messy I proposed to have a simpler solution for now which is based on what we were trying to attempt already (deferring quitting the app until it's RUNNING), as that is a single code path, so that this doesn't get much in the way of the rebase.

>If you think it's important enough to do the right thing or do nothing, then I'd suggest waiting until the rebase is done and do it afterwards. Thoughts?

I failed to properly convey my actual suggestion in my previous comment. I'm not suggesting that right now we actually kill the process without a window the moment the speedwagon is closed, even though that in theory would be the best case scenario.

What I'm suggesting is that we add logic directly in `shell-app.c` along the lines of "if we see the speedwagon window count dropping to zero, and the app would go to the STOPPED state, instead keep it in STARTING and then kill it as soon as the window appears". In other words, we'd "mask" the double transition from the outside world and completely handle internally in `shell-app.c`. Don't bother if it's too difficult though; in that case we can just merge this one.

Reply:
Well, that was what I attempted to do in the first version of this PR, maybe not in the best or most elegant way, but that was the intention: to do it self-contained in shell-app.c

That said, at that point I was still considering supporting the  -now gone- SplashClosed signal, so perhaps I can re-think it again.